internals/overlord/checkstate/manager.go

```diff
@@ -48,42 +87,114 @@ func (m *CheckManager) NotifyCheckFailed(f FailureFunc) {
 
 // PlanChanged handles updates to the plan (server configuration),
 // stopping the previous checks and starting the new ones as required.
-func (m *CheckManager) PlanChanged(p *plan.Plan) {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
+func (m *CheckManager) PlanChanged(newPlan *plan.Plan) {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// Update local reference to plan.
+	m.planLock.Lock() // always acquire locks in same order (state lock, then plan lock)
+	oldPlan := m.plan
+	m.plan = newPlan
+	m.planLock.Unlock()
+
+	if oldPlan == nil {
+		oldPlan = &plan.Plan{}
+	}
+	shouldEnsure := false
+	newOrModified := make(map[string]bool, len(newPlan.Checks))
+
+	// Abort all currently-running checks.
+	for _, change := range m.state.Changes() {
+		switch change.Kind() {
+		case performCheckKind, recoverCheckKind:
+			if change.IsReady() {
+				// Skip check changes that have finished already.
+				continue
+			}
+			details := mustGetCheckDetails(change)
+			oldConfig, inOld := oldPlan.Checks[details.Name]
+			newConfig, inNew := newPlan.Checks[details.Name]
+			if inOld && inNew {
+				if reflect.DeepEqual(oldConfig, newConfig) {
```

Comment:
Do we handle a case where the service context changes for an existing check that in itself is not modified ?

Reply:
Ah, I'd missed that! That's a really good point. And it ties into the above -- if we end up storing this on the task, It'd be better storing the final (merged-with-service-context) checker config rather just the `plan.Check`.

Reply:
Fixed now.